design/keycloak.x/operator.md

```diff
@@ -0,0 +1,123 @@
+
+### Reaugmentation process in Kubernetes
+
+We will be leveraging Kubernetes volumes to act as "caches" for the [augmented/configured](https://quarkus.io/guides/reaugmentation) version of Keycloak.
+An initial POC to show the concept has been drafted here:
+https://github.com/andreaTP/poc-mutable-jar
+
+We will use Kubernetes volumes to cache the augmented version of the binaries.
+The artifacts in the kubernetes volume will be produced by an init-container and the operation is going to result in a noop in case the volume has already been populated by a compatible augmentation.
+
+### Connecting to a database
+
+A Postgres DB instance will have to be provisioned externally, it's not Keycloak Operator's responsibility to manage a database. The DB credentials will be stored as K8s Secrets.
+
+In long-term plan we'll add a limited integration with a Postgres Operator to leverage its backup functionalities for Keycloak upgrades.
+
+### Observability
+
+The operator will provide CR metrics as well as it will provide integration with Prometheus, Grafana and AlertManager for both operator and operand metrics. This will be addressed in an upcoming design proposal.
+
+### Ingresses
+
+The operator will provide an out-of-the-box experience using an opinionated default Ingress (Route on OpenShift) configuration. This configuration will support further "manual" modification. Additionally, it will be possible to completely disable this feature.
+
+
+## Codebase
+
+The code for the new operator will be organized as a Maven sub-module in the main GitHub `keycloak/keycloak` repository.
+Dependency management will automatically piggy-back on the Keycloak BOM of the Quarkus distribution guaranteeing compliance of the used library versions.
+
+It will use the Java Operator SDK and its Quarkus extension. This implies the usage of Fabric8 K8s client.
+
+## Kubernetes compatibility matrix
+* OpenShift >=4.7
+* Vanilla Kubernetes >=1.20
+
+Other Kubernetes distributions are supported only in the best effort mode.
+
+## Distribution
+
+The Operator deployment is going to be performed leveraging OLM providing both CLI approach via `Subscription` objects for managing the operator installation, and UI in OpenShift. The Operator as such is going to be distributed as a container image.
+
+## Migration from the old operator
+No direct migration path. Generic migration steps will be documented.
+
+## Future considerations
+
+### Autonomous operator
+
+Our long-term vision includes making the operator autonomous to some extent, basically making it a [Level 5 operator](https://operatorframework.io/operator-capabilities/). It should be able to understand the operand's metrics and reflect them while automatically scaling and healing Keycloak deployment.
```

Comment:
I already mentioned this the above - have a look at KEDA.

Reply:
Thanks! Will look into it.